modules/schema/generate_test.go

```diff
@@ -89,3 +89,157 @@ creation = (D)
+	const input = `
+@url =  person
+name = "UNKNOWN" (S)    #official name of the package, there can be no overlap (can be dot notation)
+gender = "male,female,others" (E)
+	`
+
+	schema, err := New(strings.NewReader(input))
+	if err != nil {
+		panic(err)
+	}
+
+	if err := GenerateGolang(os.Stdout, "test", schema); err != nil {
+		panic(err)
+	}
+
+	// Output:
+	// package test
+	//
+	// import "encoding/json"
+	//
+	// type Person struct {
+	// 	Name   string           `json:"name"`
+	// 	Gender PersonGenderEnum `json:"gender"`
+	// }
+	//
+	// func NewPerson() (Person, error) {
+	// 	const value = "{\"name\": \"UNKNOWN\"}"
+	// 	var object Person
+	// 	if err := json.Unmarshal([]byte(value), &object); err != nil {
+	// 		return object, err
+	// 	}
+	// 	return object, nil
+	// }
+	//
+	// type PersonGenderEnum uint8
+	//
+	// const (
+	// 	PersonGenderMale PersonGenderEnum = iota
+	// 	PersonGenderFemale
+	// 	PersonGenderOthers
+	// )
+	//
+	// func (e PersonGenderEnum) String() string {
+	// 	switch e {
+	// 	case PersonGenderMale:
+	// 		return "male"
+	// 	case PersonGenderFemale:
+	// 		return "female"
+	// 	case PersonGenderOthers:
+	// 		return "others"
+	// 	}
+	// }
+}
+
+func ExampleGenerateGolang_ip() {
+	const input = `
+@url =  network
+name = "UNKNOWN" (S)    #official name of the package, there can be no overlap (can be dot notation)
+ip = "172.0.0.1" (ipaddr)
+net = "2001:db8::/32" (iprange)
+addresses = (Lipaddr)
+	`
+
+	schema, err := New(strings.NewReader(input))
+	if err != nil {
+		panic(err)
+	}
+
+	if err := GenerateGolang(os.Stdout, "test", schema); err != nil {
+		panic(err)
+	}
+
+	// Output:
+	// package test
+	//
+	// import (
+	// 	"encoding/json"
+	// 	schema "github.com/threefoldtech/zosv2/modules/schema"
+	// 	"net"
+	// )
+	//
+	// type Network struct {
+	// 	Name      string         `json:"name"`
+	// 	Ip        net.IP         `json:"ip"`
+	// 	Net       schema.IPRange `json:"net"`
+	// 	Addresses []net.IP       `json:"addresses"`
+	// }
+	//
+	// func NewNetwork() (Network, error) {
+	// 	const value = "{\"name\": \"UNKNOWN\", \"ip\": \"172.0.0.1\", \"net\": \"2001:db8::/32\"}"
+	// 	var object Network
+	// 	if err := json.Unmarshal([]byte(value), &object); err != nil {
+	// 		return object, err
+	// 	}
+	// 	return object, nil
+	// }
+}
+
+func ExampleGenerateGolang_dict() {
+	const input = `
+@url =  parent
+name = (S)    #official name of the package, there can be no overlap (can be dot notation)
+data = (dictO) ! child # dict of children
+tags = (M) # dict with no defined object type
+
+@url = child
+name = (S)
+	`
+
+	schema, err := New(strings.NewReader(input))
+	if err != nil {
+		panic(err)
+	}
+
+	if err := GenerateGolang(os.Stdout, "test", schema); err != nil {
+		panic(err)
+	}
+
+	// Output:
+	// package test
+	//
+	// import "encoding/json"
+	//
+	// type Parent struct {
+	// 	Name string                 `json:"name"`
+	// 	Data map[string]Child       `json:"data"`
```

Comment:
I'm not sure how JSX handles this, but does dict only allows string as key ?

Reply:
In Python any hashable object can be used as a key. In Go anything can be used as key i guess maybe except empty struct. I was asking regarding the implementation of schema in jsx. I think they only allow string l. Because the dict schema definition is onlyd (dict) not followed by element type. Supporting an element type (like what list do) i added to the go implementation.

Reply:
> In Go anything can be used as key 

I double checked that and this is actual not completely correct, see: 

> The comparison operators == and != must be fully defined for operands of the key type;

from https://golang.org/ref/spec#Map_types

I also did the check in JSX itself:
```python
SCHEMA = """ 
@url = test.schema.1 
info = (dict) 
init_dict = {"number": 468} (dict) 
""" 
schema = j.data.schema.get_from_text(SCHEMA)                                                                                                                                                                                              
obj = schema.new()                                                                                                                                                                                                                        
obj.info = {1:'hello',"world":2}                                                                                                                                                                                                          
{1: 'hello', 'world': 2}
obj._json                                                                                                                                                                                                                                 
'{\n"info": {\n"1": "hello",\n"world": 2\n},\n"init_dict": {\n"number": 468\n}\n}'
```
So it seems JSX schema use a plain python dict so any hashable value can be used as a key. Even wrost for us, this mean we can have mix type in the keys.

I think we should clearly define our limitation in the go implementation and warn the user when such data is received. So it will be the responsibility to the user to make sure to not go too crazy with these things when implementing something that needs to be use by our go implementation.

Reply:
> I double checked that and this is actual not completely correct.
Yup, you are right, i completely forgot about the `==` and `!=` implementation.

I totally understand that we can't support everything that python do. I suggest an extension to the schema language as follows (I can implement this easily in Go):

For dict type, the Key, and Value types can be defined:
- `dict` : default to `Key == String`, and `Value == interface{}`
- `dictT` default to `Key == String` and defines `Value == T`
- `dictKV` defines `Key == K` and `Value == V`

I think this can be easily implemented in `python` as for Go. I also suggest that python add a short name for dict (as `M` since `D` is already taken by `Date`)

Reply:
Looks good. We don't really need dict at the moment so I won't block the PR on this. But this solution could work if we ever need to use dict